### PR TITLE
Hw33

### PR DESCRIPTION
**Pull request by pavlokhomiak**

### CODE REVIEW

**Review by tarkaiev — COMMENTED** [review]
did you try to remove product from shopping cart?)

File: src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                PreparedStatement insertOrderQueryStatement
+                        = connection.prepareStatement(insertOrder,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            insertOrderQueryStatement.setLong(1, order.getUserId());
+            insertOrderQueryStatement.executeUpdate();
+            ResultSet resultSet = insertOrderQueryStatement.getGeneratedKeys();
+            if (resultSet.next()) {
+                Long orderId = resultSet.getLong(1);
+                order.setId(orderId);
+                insertOrdersProducts(order, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java:35]: it's better not to open 2 statements at the same time, you can think of the way to close previous one before opening new in insertOrdersProducts

File: src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    @Override
+    public Order update(Order order) {
+        String updateOrder = "UPDATE orders SET user_id = ? WHERE order_id = ? AND deleted = false";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement updateOrderStatement
+                        = connection.prepareStatement(updateOrder)) {
+            updateOrderStatement.setLong(1, order.getUserId());
+            updateOrderStatement.setLong(2, order.getId());
+            if (updateOrderStatement.executeUpdate() == 1) {
+                updateOrdersProducts(order, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java:54]: you open three statements at the same time here
check all method, where you can have few connections/statements at the same time and fix it

File: src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        String updateOrder = "UPDATE orders SET deleted = true WHERE order_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement updateOrderStatement
+                        = connection.prepareStatement(updateOrder);) {
+
+            updateOrderStatement.setLong(1, id);
+            if (updateOrderStatement.executeUpdate() == 0) {
+                return false;
+            }
+            deleteOrdersProducts(id, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java]: don't delete order products, marking order as deleted is enough

File: src/main/java/mate/academy/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                PreparedStatement insertShoppingCartsStatement
+                        = connection.prepareStatement(insertShoppingCarts,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            insertShoppingCartsStatement.setLong(1, shoppingCart.getUserId());
+            insertShoppingCartsStatement.executeUpdate();
+            ResultSet resultSet = insertShoppingCartsStatement.getGeneratedKeys();
+            if (resultSet.next()) {
+                Long shoppingCartId = resultSet.getLong(1);
+                shoppingCart.setId(shoppingCartId);
+                insertShoppingCartsProducts(shoppingCart, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/ShoppingCartDaoJdbcImpl.java:35]: same with statements in this class

File: src/main/java/mate/academy/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    private void insertShoppingCartsProducts(ShoppingCart shoppingCart, Connection connection)
+            throws SQLException {
+        String insertShoppingCartsProducts
+                = "INSERT INTO shopping_carts_products (cart_id, product_id) VALUES (?, ?);";
+        try (PreparedStatement insertStatement
+                     = connection.prepareStatement(insertShoppingCartsProducts)) {
+            for (Product product : shoppingCart.getProducts()) {
+                setProductId(product, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: in update method you should already have products with ids, no need to fetcher them additionally here

File: src/main/java/mate/academy/dao/jdbc/UserDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            if (resultSet.next()) {
+                Long userId = resultSet.getLong(1);
+                user.setId(userId);
+                insertUserRoles(user, connection);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Creating user " + user
+                    + " is failed", e);
+        }
+        return user;
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/UserDaoJdbcImpl.java:44]: does user that you return has ids for his roles, if yes, when did you set them ?

File: src/main/java/mate/academy/dao/jdbc/UserDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        String updateUsers = "UPDATE users SET deleted = true WHERE user_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement updateUsersStatement
+                        = connection.prepareStatement(updateUsers)) {
+
+            updateUsersStatement.setLong(1, id);
+            if (updateUsersStatement.executeUpdate() == 1) {
+                return false;
+            }
+            deleteUsersRoles(id, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/academy/dao/jdbc/UserDaoJdbcImpl.java]: don't delete roles, just mark user as deleted

File: src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+    @Override
+    public Order update(Order order) {
+        String updateOrder = "UPDATE orders SET user_id = ? WHERE order_id = ? AND deleted = false";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement updateOrderStatement
+                        = connection.prepareStatement(updateOrder)) {
+            updateOrderStatement.setLong(1, order.getUserId());
+            updateOrderStatement.setLong(2, order.getId());
+            if (updateOrderStatement.executeUpdate() == 1) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/dao/jdbc/OrderDaoJdbcImpl.java:52]: if user_id has not been changed will the update method return 1? If yes then it's okay, if not then reconsider this solution

File: src/main/java/mate/academy/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    private void insertShoppingCartsProducts(ShoppingCart shoppingCart, Connection connection)
+            throws SQLException {
+        String insertShoppingCartsProducts
+                = "INSERT INTO shopping_carts_products (cart_id, product_id) VALUES (?, ?);";
+        try (PreparedStatement insertStatement
+                     = connection.prepareStatement(insertShoppingCartsProducts)) {
+            for (Product product : shoppingCart.getProducts()) {
+                if (product.getId() == null) {
+                    setProductId(product, connection);
+                }
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: do we really need this check?

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**